sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,555 @@
+
+  /**
+   * Returns an option of Statistics for a Filter logical plan node.
+   * For a given compound expression condition, this method computes filter selectivity
+   * (or the percentage of rows meeting the filter condition), which
+   * is used to compute row count, size in bytes, and the updated statistics after a given
+   * predicated is applied.
+   *
+   * @param plan a LogicalPlan node that must be an instance of Filter.
+   * @return Option[Statistics] When there is no statistics collected, it returns None.
+   */
+  def estimate(plan: Filter): Option[Statistics] = {
+    val stats: Statistics = plan.child.statistics
+    if (stats.rowCount.isEmpty) return None
+
+    // save a mutable copy of colStats so that we can later change it recursively
+    val statsExprIdMap: Map[ExprId, ColumnStat] =
+      stats.attributeStats.map(kv => (kv._1.exprId, kv._2))
+    mutableColStats = mutable.Map.empty ++= statsExprIdMap
+
+    // estimate selectivity of this filter predicate
+    val percent: Double = calculateConditions(plan, plan.condition)
+
+    // attributeStats has mapping Attribute-to-ColumnStat.
+    // mutableColStats has mapping ExprId-to-ColumnStat.
+    // We use an ExprId-to-Attribute map to facilitate the mapping Attribute-to-ColumnStat
+    val expridToAttrMap: Map[ExprId, Attribute] =
+      stats.attributeStats.map(kv => (kv._1.exprId, kv._1))
+    // copy mutableColStats contents to an immutable AttributeMap.
+    val mutableAttributeStats: mutable.Map[Attribute, ColumnStat] =
+      mutableColStats.map(kv => expridToAttrMap(kv._1) -> kv._2)
+    val newColStats = AttributeMap(mutableAttributeStats.toSeq)
+
+    val filteredRowCountValue: BigInt =
+      EstimationUtils.ceil(BigDecimal(stats.rowCount.get) * percent)
+    val avgRowSize = BigDecimal(EstimationUtils.getRowSize(plan.output, newColStats))
+    val filteredSizeInBytes: BigInt =
+      EstimationUtils.ceil(BigDecimal(filteredRowCountValue) * avgRowSize)
+
+    Some(stats.copy(sizeInBytes = filteredSizeInBytes, rowCount = Some(filteredRowCountValue),
+      attributeStats = newColStats))
+  }
+
+  /**
+   * Returns a percentage of rows meeting a compound condition in Filter node.
+   * A compound condition is depomposed into multiple single conditions linked with AND, OR, NOT.
+   * For logical AND conditions, we need to update stats after a condition estimation
+   * so that the stats will be more accurate for subsequent estimation.  This is needed for
+   * range condition such as (c > 40 AND c <= 50)
+   * For logical OR conditions, we do not update stats after a condition estimation.
+   *
+   * @param plan the Filter LogicalPlan node
+   * @param condition the compound logical expression
+   * @param update a boolean flag to specify if we need to update ColumnStat of a column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   */
+  def calculateConditions(

```

Comment:
I hesitate to make it private because we may extend this class in the future.  Do you remember what you once said that a new approach is to sample records uniformly?  Instead of using histogram (or count min sketch), for any selectivity estimate, we may always run a query against the small sample set to estimate.  If we want do so, then we can extend FilterEstimation class and override  calculateConditions method.